File: MazeMan.java
import java.util.Scanner;

public class MazeMan {
    private static int dots;

    private static void explore(char[][] maze, int i, int j) {
        int n = maze.length, m = maze[0].length;
        if (maze[i][j] == '.') dots++;
        maze[i][j] = 'X';

        // north
        if (i > 0 && (maze[i - 1][j] == ' ' || maze[i - 1][j] == '.'))
            explore(maze, i - 1, j);

        // south
        if (i < n - 1 && (maze[i + 1][j] == ' ' || maze[i + 1][j] == '.'))
            explore(maze, i + 1, j);
        
        // east
        if (j < m - 1 && (maze[i][j + 1] == ' ' || maze[i][j + 1] == '.'))
            explore(maze, i, j + 1);
        
        // west
        if (j > 0 && (maze[i][j - 1] == ' ' || maze[i][j - 1] == '.'))
            explore(maze, i, j - 1);
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt(), m = sc.nextInt();
        sc.nextLine();
        char[][] maze = new char[n][];
        for (int i = 0; i < n; i++) maze[i] = sc.nextLine().toCharArray();
        
        int players = 0;

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (maze[i][j] >= 'A' && maze[i][j] < 'X') {
                    dots = 0;
                    explore(maze, i, j);
                    if (dots > 0) players++;
                }
            }
        }

        dots = 0;

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (maze[i][j] == '.') dots++;
            }
        }

        System.out.println(players);
        System.out.println(dots);
    }
}

File: maze_man.py
# Based on Java version, which works, but fails later test case
# with a run time error.

def explore(maze, i, j):
    global dots

    n, m = len(maze), len(maze[0])
    if maze[i][j] == '.': dots += 1
    maze[i][j] = 'X'

    # north
    if i > 0 and (maze[i - 1][j] == ' ' or maze[i - 1][j] == '.'):
        explore(maze, i - 1, j)
    
    # south
    if i < n - 1 and (maze[i + 1][j] == ' ' or maze[i + 1][j] == '.'):
        explore(maze, i + 1, j)

    # east
    if j < m - 1 and (maze[i][j + 1] == ' ' or maze[i][j + 1] == '.'):
        explore(maze, i, j + 1)

    # west
    if j > 0 and (maze[i][j - 1] == ' ' or maze[i][j - 1] == '.'):
        explore(maze, i, j - 1)

if __name__ == '__main__':
    global dots

    n, m = [int(i) for i in input().split()]
    maze = [list(input()) for i in range(n)]
    players = 0

    for i in range(n):
        for j in range(m):
            if maze[i][j] >= 'A' and maze[i][j] < 'X':
                dots = 0
                explore(maze, i, j)
                if dots > 0: players += 1

    dots = 0

    for i in range(n):
        for j in range(m):
            if maze[i][j] == '.': dots += 1

    print(players)
    print(dots)